### ISSUE TITLE
[Clang]: Incorrect inheritance of protected (default) constructor

### ISSUE THREAD

**Issue by Krzmbrzl** [llvm/llvm-project#59996]
The snippet
```cpp
#include <iostream>
#include <type_traits>

class Foo {
protected:
   Foo() = default; 
};

class Bar : public Foo {
public:
    using Foo::Foo;

    Bar(const Bar &) = default;
};

int main() {
    std::cout << std::boolalpha << "Is default constructible: "
        << std::is_default_constructible_v< Bar > << std::endl;
    
    Bar b;
}
```
compiles fines with Clang (trunk and older), whereas it fails to compile with GCC and MSVC because `Bar` has no public default constructor.

If the line `Bar b;` is commented out, the program compiled with Clang will report `true` whereas GCC and MSVC report `false`. I suspect that Clang's behavior here is related to the abovementioned problem.

Note: If the visibility of `Bar`'s constructor is changed to `private`, then everything works as expected.

[Godbolt](https://godbolt.org/z/ex3eWoh5z)

I am not nearly competent enough to look up the expected behavior in the C++ standard, but given that GCC  and MSVC seem to agree on what the result should be, I figured that this is likely a bug in Clang :shrug:

**Comment by zygoloid** [comment]
I don't think this is an access problem. In the AST dump it looks like clang is declaring an implicit default constructor for `Bar`, even though `Bar` explicitly declares a constructor, which should inhibit the implicit declaration of the default constructor. (And the implicit default constructor for `Bar` is valid because it's permitted to call a protected default constructor of the base class.)
